### PR TITLE
Reduce the number of docvalues iterator created in the global ordinals fielddata

### PR DESCRIPTION
**Pull request by jimczi**
Today the fielddata for global ordinals re-creates docvalues readers of each segment
when building the iterator of a single segment. This is required because the lookup of
global ordinals needs to access the docvalues's TermsEnum of each segment to retrieve
the original terms. 
This also means that we need to create NxN (where N is the number of segment in the index) docvalues iterators each time we want to collect global ordinal values. 
This wasn't an issue in previous versions since docvalues readers are stateless before 6.0 so they are reused on each segment but now that docvalues are iterators (>= ES 6.0) we need to create a new instance each time we want to access the values. 
In order to avoid creating too many iterators this change splits the global ordinals fielddata in two classes, one that is used to cache a single instance per directory reader and one
that is created from the cached instance that can be used by a single consumer. The latter creates the TermsEnum of each segment once and reuse them to create the segment's iterator. This prevents the creation of all TermsEnums each time we want to access the value of a single segment, hence reducing the number of docvalues iterator to create to Nx2 (one iterator and one lookup per segment).

### CODE REVIEW

**Comment by jtibshirani** [comment]
I am working to get up to speed on the code and wanted to check I understood the motivation for adding a new inner class `Consumer`. An alternative approach would have been to keep the current structure `GlobalOrdinalsIndexFieldData`, and create the `TermsEnum` lookup instances once inside of the `GlobalOrdinalsIndexFieldData` constructor, then pass these lookup instances when creating the `Atomic` readers. But I don't think this works because the docvalues can't be created in advance like that (from a potentially different thread?)

In this PR, `GlobalOrdinalsIndexFieldData` no longer really functions as field data, but just as an object to hold cached information. I wonder if it'd be possible to refactor it to make this clearer, as this would let us avoid the unsupported methods ("loadDirect(LeafReaderContext) should not be called in this context").

**Comment by jimczi** [comment]
> But I don't think this works because the docvalues can't be created in advance like that (from a potentially different thread?)

Yes, that's the issue. `TermsEnum`s are not thread safe so the other alternative is to use a `ThreadLocal` but after discussing with @jpountz we agreed that we should try to avoid this path. The approach here still uses a single cached instance and creates a new copy of it each time a consumer needs to access the global ordinals (`loadGlobal`). This way we can reuse the lookups (the `TermsEnum`) on each segment, hence avoiding the NxN explosion of the current code.

File: server/src/main/java/org/elasticsearch/index/fielddata/ordinals/GlobalOrdinalsIndexFieldData.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/fielddata/ordinals/GlobalOrdinalsIndexFieldData.java]: Small comment, I think this can be package-private now that the class is marked as `final`.

**Review by jtibshirani — APPROVED** [review]
This looks good to me, it seems like the only viable approach given how the global ordinals/ field data framework is currently designed. I'll approve, but understand if you want a second opinion since I'm not an expert in this part of the code.

The global ordinals code is already complex, and this change adds more complexity, so it would be really good to refactor it soon. I will try to help come up with ideas for simplifying it.

**Review by jpountz — APPROVED** [review]
A bit hacky, but I don't have a better suggestion and it fixes an important performance bug.

**Pull request merged**
